File: KP_Methods/C++/Brute_Force.cpp
#include <iostream>
#include <vector>
using namespace std;

int get_weight(vector<int> item, vector<int> elements) {
    int weight = 0;
    for (int i = 0; i < elements.size(); i++) {
        if (item[i] == 1) {
            weight += elements[i];
        }
    }
    return weight;
}

vector<vector<int>> subsets(vector<vector<int>> items) {
    vector<vector<int>> subsets = {{}};
    for (vector<int> item : items) {
        int n = subsets.size();
        for (int i = 0; i < n; i++) {
            vector<int> new_subset;
            for (int j = 0; j < subsets[i].size(); j++) {
                new_subset.push_back(subsets[i][j]);
            }
            for (int j = 0; j < item.size(); j++) {
                new_subset.push_back(item[j]);
            }
            subsets.push_back(new_subset);
        }
    }
    return subsets;
}

int knapsack_bruteForce(vector<int> elements, vector<vector<int>> items, vector<int> profits, int max_weight) {
    // Helper function to generate all possible subsets of items
    vector<vector<int>> all_subsets = subsets(items);

    // Initialize the maximum profit and the best subset of items
    int max_profit = 0;
    vector<int> best_subset;

    for (vector<int> subset : all_subsets) {
        if (get_weight(subset, elements) <= max_weight) {
            int subset_profit = 0;
            for (int i = 0; i < subset.size(); i++) {
                if (subset[i] == 1) {
                    subset_profit += profits[i];
                }
            }
            if (subset_profit > max_profit) {
                max_profit = subset_profit;
                best_subset = subset;
            }
        }
    }
    return max_profit;
}

int main() {
    vector<int> elements = {10, 20, 30};
    vector<vector<int>> items = {{1, 1, 0}, {1, 0, 1}, {0, 1, 1}};
    vector<int> profits = {60, 100, 120};
    int max_weight = 50;
    cout << knapsack_bruteForce(elements, items, profits, max_weight) << endl;
    return 0;
}

File: KP_Methods/Python/Brute_Force.py
def knapsack_bruteForce(elements, items, profits, max_weight):
    # Helper function to calculate the weight of an item
    def get_weight(item):
        weight = 0
        for i in range(len(item)):
            if item[i] == 1:
                weight += elements[i]
        return weight

    # Helper function to generate all possible subsets of items
    def subsets(items):
        subsets = [[]]
        for item in items:
            for i in range(len(subsets)):
                subsets.append(subsets[i] + [item])
        return subsets

    # Initialize the maximum profit and the best subset of items
    max_profit = 0
    best_subset = []
    all_subsets = subsets(range(len(items)))
    for subset in all_subsets:
        subset_items = [items[i] for i in subset]
        if get_weight(subset_items) <= max_weight:
            subset_profit = sum([profits[i] for i in subset])
            if subset_profit > max_profit:
                max_profit = subset_profit
                best_subset = subset_items
    return max_profit

# Example usage
elements = [10, 20, 30]
items = [[1, 1, 0], [1, 0, 1], [0, 1, 1]]
profits = [60, 100, 120]
max_weight = 50
print(knapsack_bruteForce(elements, items, profits, max_weight))

import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'
import numpy as np

# Selection of instance
data = pd.read_csv(r'C:\Users\Javier\Documents\IA\Fondecyt\KP_Methods\Instances\InstancesofSetI\85_100_0.10_0.75.txt', header=None)

n_items = data[0][0]
n_elements = data[0][1]
max_weight = int(data[0][2])

profits = str(data[0][3]).split()
for i in range(len(profits)):
    profits[i] = int(profits[i])

elements = str(data[0][4]).split()
for i in range(len(elements)):
    elements[i] = int(elements[i])

items = data.tail(-5)
items.reset_index(inplace = True, drop = True)

for i in range(len(items)):
    items[0][i] = str(items[0][i]).split()
for i in range(len(items)):
    for j in range(len(elements)):
        items[0][i][j] = int(items[0][i][j])

items = items[0].values.tolist()

print(knapsack_bruteForce(elements, items, profits, max_weight))
